pint_server/app.py

```diff
@@ -306,70 +452,178 @@ def _get_azure_images_for_region_state(region, state):
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].region == region,
+            PROVIDER_IMAGES_MODEL_MAP[provider].state == ImageState.deprecated,
+            PROVIDER_IMAGES_MODEL_MAP[provider].deprecatedon < deprecatedby,
+        )
+
+    return images
+
+
+def get_provider_images_to_be_deletedby(deletedby, provider, region=None):
+    images = query_deletedby_images_in_provider_region(
+        deletedby, provider, region)
+
+    extra_attrs = {}
+    if region and provider == 'microsoft':
+        extra_attrs['region'] = region
+
+    return formatted_provider_images(provider, images,
+                                     extra_attrs=extra_attrs)
+
+
+def _query_image_in_azure_region(image_name, region):
+    # lookup environment for given region, assuming unique per region
+    environment_name = _get_azure_environment_name_for_region(region)
+
+    # retrieve matching images for region
+    images = MicrosoftImagesModel.query.filter(
+        MicrosoftImagesModel.environment == environment_name,
+        MicrosoftImagesModel.name == image_name)
+
+    return images
+
+
+def query_image_in_provider_region(image_name, provider, region=None):
+    if region is None:
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].name == image_name)
+    # microsoft needs special handling for region queries
+    elif provider == 'microsoft':
+        images = _query_image_in_azure_region(image_name, region)
+
+    # if provider has regions retrieve matching images for region
+    elif (hasattr(PROVIDER_IMAGES_MODEL_MAP[provider], 'region')):
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].region == region,
+            PROVIDER_IMAGES_MODEL_MAP[provider].name == image_name)
+
+    return images
+
+
+def get_image_deletiondate_in_provider(image, provider, region=None):
+
+    images = query_image_in_provider_region(image, provider, region)
+
+    # depending on provider there may be multiple images all of
+    # which should have the same state, deprecatedon and deletedon
+    # values. Use a set to eliminate duplicates, and a namedtuple
+    # helper to create a hashable object.
+    image_set = {DeletionDetails(i.state,
+                                 i.deprecatedon,
+                                 i.deletedon)
+                 for i in images
+                 if i.state in [ImageState.deprecated,
+                                ImageState.deleted]}
+
+    # TODO(rtamalin): is this even possible?
+    if len(image_set) > 1:
         abort(Response('', status=404))
-    exclude_attrs = PROVIDER_IMAGES_EXCLUDE_ATTRS.get(provider)
-    return [get_formatted_dict(image, exclude_attrs=exclude_attrs)
-            for image in images]
 
+    # if image is in not in deprecated/deleted state the image set
```

Comment:
s/in//

Reply:
My understanding is that the lambda function instance sticks around for a few mins before being disposed of, so if subsequent queries arrive during that timeframe then then could potentially benefit from the cached values. If there is concern about the size of the cached values (which isn't necessarily too large right now) we could reduce the cache size down to 1 so that the caching benefit is primarily focused the current request, and only benefits subsequent requests if they are for the same provider.